nixos/modules/system/boot/initrd-openvpn.nix

```diff
@@ -0,0 +1,112 @@
+        included. Service is killed when stage-1 boot is finished.
+      '';
+    };
+    
+    boot.initrd.network.openvpn.files = mkOption {
+      default = {};
+      example = ''
+        {
+          "tls.key" = /root/tls.key;
+          "ca.crt" = /root/ca.crt;
+        }
+      '';
+      description = ''
+        Additional files to use in the configuration. These are stored in 
+        <filename>/etc/openvpn/</filename>. Warning: Unless your bootloader 
+        supports initrd secrets, these are insecurely added to the global Nix 
+        store.
+      '';
+      
+      type = types.attrsOf (types.path);
+    };
+    
+    boot.initrd.network.openvpn.configuration = mkOption {
+      type = with types; nullOr (either lines path);
+      default = null;
+      description = ''
+        The configuration file for OpenVPN. Warning: This is insecurely added to
+        the global Nix store if a string is used for this option.
+        If a path is used, it depends on the support of your bootloader for 
+        initrd secrets. If these are unsupported, the configuration will be 
+        added insecurely to the global Nix store.
+        
+        Additionally, if a path is used, the configuration has to include the
+        option <literal>iproute /bin/ip</literal>, as openvpn is not able to
+        use the correct binary otherwise.
+      '';
+    };
+
+  };
+
+  config = mkIf (config.boot.initrd.network.enable && cfg.enable) {
+    assertions = [
+      { assertion = cfg.configuration != null;
+        message = "You should specify a configuration for the initrd OpenVPN.";
+      }
+    ];
+
+    # Add kernel modules needed for OpenVPN
+    boot.initrd.kernelModules = [ "tun" "tap" ];
+
+    # Add custom files and configuration to the initrd
+    # For additional files, it is bypassing the nix store if initrd secrets are
+    # available.
+    # For the configuration, it abuses boot.initrd.secrets to conveniently
+    # add the file (and bypasses the store if a path is used and initrd secrets
+    # are available).
+    boot.initrd.secrets = mapAttrs'
+                            (name: value:
+                              nameValuePair ("/etc/openvpn/" + name) (value))
+                            (cfg.files //
+                              {"ovpn-initrd-cfg.ovpn" = cfgFile;});
+
+    # Add openvpn and ip binaries to the initrd
+    # The shared libraries are required for DNS resolution
+    boot.initrd.extraUtilsCommands = ''
+      copy_bin_and_libs ${pkgs.openvpn}/bin/openvpn
+      copy_bin_and_libs ${pkgs.iproute}/bin/ip
+
+      cp -pv ${pkgs.glibc}/lib/libresolv.so.2 $out/lib
+      cp -pv ${pkgs.glibc}/lib/libnss_dns.so.2 $out/lib
+    '';
+    
+    # openvpn --version would exit with 1 instead of 0
+    boot.initrd.extraUtilsCommandsTest = ''
+      $out/bin/openvpn --show-gateway 
+    '';
+
+    boot.initrd.network.postCommands = ''
+      openvpn /etc/openvpn/ovpn-initrd-cfg.ovpn &
```

Comment:
Don't we have systemd in the initrd?

Reply:
I've only seen `systemd-udevd`, `libsystemd-shared-239.so` and `libsystemd.so.0` in the initramfs, and as `initrd-ssh.nix` uses the same approach to start `dropbear` ([see here](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/initrd-ssh.nix#L122)), I assumed that this is the correct way of starting the client.
Additionally, this gives us the output of the client on the tty, which can make configuration/connection problems more obvious (even when e.g. restricting the shell to `/bin/cryptsetup-askpass`, what I guess most people do).

Reply:
No we don't